lib/pbench/server/api/__init__.py

```diff
@@ -333,53 +336,82 @@ def post(self):
+            app.logger.debug("Writing chunks")
+            hash_md5 = hashlib.md5()
 
+            try:
                 while True:
                     chunk = request.stream.read(chunk_size)
                     bytes_received += len(chunk)
                     if len(chunk) == 0 or bytes_received > content_length:
                         break
 
-                    f.write(chunk)
+                    ofp.write(chunk)
                     hash_md5.update(chunk)
-        except Exception:
-            app.logger.exception(
-                "Tarfile upload: There was something wrong uploading {}", filename
-            )
-            abort(500, message=f"There was something wrong uploading {filename}")
+            except Exception:
+                app.logger.exception(
+                    "Tarfile upload: There was something wrong uploading {}", filename
+                )
+                abort(500, message=f"There was something wrong uploading {filename}")
 
-        message = None
-        if bytes_received != content_length:
-            app.logger.debug(
-                f"Tarfile upload: Bytes received does not match with content length header value for file {filename}"
-            )
-            message = (
-                f"Bytes received ({bytes_received}) does not match with content length header"
-                f" ({content_length}), upload failed"
-            )
+            if bytes_received != content_length:
+                app.logger.debug(
+                    f"Tarfile upload: Bytes received does not match with content length header value for file {filename}"
+                )
+                message = (
+                    f"Bytes received ({bytes_received}) does not match with content length header"
+                    f" ({content_length}), upload failed"
+                )
+                abort(400, message=message)
 
-        elif hash_md5.hexdigest() != md5sum:
-            app.logger.debug(f"Tarfile upload: md5sum check failed for file {filename}")
-            message = f"md5sum check failed for {filename}, upload failed"
+            elif hash_md5.hexdigest() != md5sum:
+                app.logger.debug(
+                    f"Tarfile upload: md5sum check failed for file {filename}"
+                )
+                message = f"md5sum check failed for {filename}, upload failed"
+                abort(400, message=message)
 
-        if message:
+            # First write the .md5
             try:
-                os.remove(full_path)
-            except Exception as exc:
-                app.logger.warning(
-                    "Failed to remove compressed %s when trying to clean up: %s",
-                    full_path,
-                    exc,
+                with md5_full_path.open("w") as md5fp:
+                    md5fp.write(f"{md5sum} {filename}\n")
+            except Exception:
+                try:
+                    os.remove(md5_full_path)
+                except Exception as exc:
```

Comment:
We should ignore `FileNotFoundError` exceptions since the creation of the `.md5` file might not have happened.